### ISSUE TITLE
Resolve local hostnames

### ISSUE THREAD

**Issue by F0cks** [pivpn/pivpn#549]
Hello guys!
I would like to know if there is a way to resolve local hostnames?
For example I have a computer with the hostname _test_ and with the IP: _192.168.0.2_.
On my network (from an other computer) I can ping `test.local` and `192.168.0.2`.
Now if I connect to my network through PiVPN (with Ubuntu 16.04.4 and OpenVPN OpenVPN 2.3.10 or Iphone APP), I can only ping `192.168.0.2` and not `test.local`.
Do I need a to set up a DNS server on my PiVPN server or sometihing in _server.conf_ can do the trick?
Something like `push "dhcp-option DNS *some*IP"`
Thanks for your help and your amazing work!
You made vpn server really easy to create!

Edit: 
I have tried (as suggested here [#265](https://github.com/pivpn/pivpn/issues/265#issuecomment-312812537)) to change dns ip with my router ip:
```
pi@vpn:~ $ ip route | grep default
default via 192.168.0.254 dev eth0 src 192.168.0.22 metric 202
```

```
File: /etc/openvpn/server.conf                      

#push "dhcp-option DNS 208.67.222.222"
#push "dhcp-option DNS 208.67.220.220"
push "dhcp-option DNS 192.168.0.254"
```
But it doesn't work:
```
ping test.local
ping: unknown host test.local

ping 192.168.0.2
PING 192.168.0.2 (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=64 time=84.7 ms
64 bytes from 192.168.0.2: icmp_seq=2 ttl=64 time=78.2 ms
64 bytes from 192.168.0.2: icmp_seq=3 ttl=64 time=76.5 ms
^C
--- 192.168.0.2 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 76.547/79.836/84.738/3.547 ms
```

**Comment by orazioedoardo** [comment]
Did you restart the server and reconnected after the test? Run `sudo systemctl restart openvpn` on the raspberry pi.

**Comment by F0cks** [comment]
Thanks for you answer @orazioedoardo !
I have tried that, I have also tried to reboot and I have tried :
`sudo systemctl stop openvpn@server; sudo systemctl enable openvpn@server.service; sudo systemctl start openvpn@server`
(Just to be sure that _restart_ was not the problem :D) 
It seems to be a common problem with openvpn servers, but all solutions that I have found not work for now ....

**Comment by cfcolaco** [comment]
Hi @F0cks 

there are several resolved / clsoed issues in this repository that address that situation and a search could have spared you the waiting time.
 
The easy and less painful way to maintain is Instead of using Google or whatever DNS you are using in your openvpn server you need to set the DNS to go to your router. 

so edit /etc/openvpn/server.conf and change the ip in 'push "dhcp-option DNS 8.8.8.8"' to match the ip address of your router, restart openvpn service 'sudo systemctl restart openvpn' , and all should be working. 

another alternative is edit /etc/hosts on your raspberry pi and add 

'<IPAddress> <hostname>'

to that file. 
will add this information to the wiki to make it easier for future.

**Comment by F0cks** [comment]
Hello @cfcolaco and thanks for your reply
I have made that research then I have posted this question because none of the solution that I have found in issues has worked.
As I have written to @orazioedoardo, I have already tried to replace OpenDNS ip in server.conf with my router ip. 
Edit hosts doesn't seem to be an elegant solution because it requires static ips the on network. But why not if it works. 
So here is my new hosts file:
```
GNU nano 2.7.4                   File: /etc/hosts                             

127.0.0.1       localhost
::1             localhost ip6-localhost ip6-loopback
ff02::1         ip6-allnodes
ff02::2         ip6-allrouters

192.168.0.2    test
```
I have rebooted the system and btw, I have kept the router ip in server.conf:
```
GNU nano 2.7.4            File: /etc/openvpn/server.conf                      

[...]
#push "dhcp-option DNS 8.8.8.8"
push "route 192.168.0.254"
[...]
```
I have used nmap to see if the vpn server was able to resolve hostnames on my network:
```
sudo nmap -sn 192.168.0.0/24

Starting Nmap 7.40 ( https://nmap.org ) at 2018-06-25 06:56 UTC
Nmap scan report for test 192.168.0.2  <- Hostname  found: test
Host is up (0.0055s latency).
MAC Address: XX:XX:XX:XX:XX:XX (Freebox SAS)
Nmap scan report for 192.168.0.254
Host is up (0.00051s latency).
MAC Address: XX:XX:XX:XX:XX:XX (Unknown)
Nmap scan report for vpn (192.168.0.22) <- Hostname found: vpn
```
So the vpn sees its own hostname "vpn" and my target "test"
Then I have run the same command on a computer connected to the vpn:
```
sudo nmap -sn 192.168.0.0/24

Starting Nmap 7.01 ( https://nmap.org ) at 2018-06-25 08:44 CEST
Nmap scan report for 192.168.0.22 <- Hostname not found
Host is up (0.075s latency).
Nmap scan report for 192.168.0.49
Host is up (0.15s latency).
Nmap scan report for 192.168.0.2 <- Hostname not found
Host is up (0.25s latency).
```
Computer is still not able to see distant hostnames...

> will add this information to the wiki to make it easier for future.

Maybe not yet because it doesn't seem to work for everyone... Like me ... :'(

**Comment by orazioedoardo** [comment]
What’s the content of `/etc/resolv.conf` on the VPN client (after you connect to the raspberry)?

**Comment by cfcolaco** [comment]
> GNU nano 2.7.4            File: /etc/openvpn/server.conf                      
>
>[...]
>#push "dhcp-option DNS 8.8.8.8"
>push "route 192.168.0.254"           <----  router ip
>[...]

Can you see what is wrong there? =) 

`push "route 192.168.0.254"` is not the same thing as push `"dhcp-options DNS 192.168.0.254"`

Can you please provide the full content ( hiding Public IP address), of your /etc/openvpn/server.conf ??

**Comment by F0cks** [comment]
> Can you see what is wrong there? =)

Damn you are right @cfcolaco ! Sadly, it is just a copy/paste error... (I have tried _route_ and _dhcp-options_ together just for testing...) 
Here is my _server.conf_:
```
File: /etc/openvpn/server.conf                      

dev tun
proto udp
port XXXX
ca /etc/openvpn/easy-rsa/pki/ca.crt
cert /etc/openvpn/easy-rsa/pki/issued/server_XXX.crt
key /etc/openvpn/easy-rsa/pki/private/server_XXX.key
dh /etc/openvpn/easy-rsa/pki/dh2048.pem
topology subnet
server 10.8.0.0 255.255.255.0
# Set your primary domain name server address for clients
#push "dhcp-option DNS 8.8.8.8"
push "dhcp-option DNS 192.168.0.254"
# Prevent DNS leaks on Windows
push "block-outside-dns"
# Override the Client default gateway by using 0.0.0.0/1 and
# 128.0.0.0/1 rather than 0.0.0.0/0. This has the benefit of
# overriding but not wiping out the original default gateway.
push "redirect-gateway def1"
client-to-client
keepalive 1800 3600
remote-cert-tls client
tls-version-min 1.2
tls-auth /etc/openvpn/easy-rsa/pki/ta.key 0
cipher AES-256-CBC
auth SHA256
comp-lzo
user nobody
group nogroup
persist-key
persist-tun
crl-verify /etc/openvpn/crl.pem
status /var/log/openvpn-status.log 20
status-version 3
syslog
verb 3
#DuplicateCNs allow access control on a less-granular, per user basis.
#Remove # if you will manage access by user instead of device.
#duplicate-cn
# Generated for use by PiVPN.io
```
And as @orazioedoardo asked me, here is my _resolv.conf_ file on the client computer (connected to the vpn) (Ubuntu 16.04.4 and OpenVPN OpenVPN 2.3.10):
```
File: /etc/resolv.conf                              

# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 127.0.1.1
search lan
```
But I can't reach hostnames on my Iphone with openvpn app neither.

**Comment by cfcolaco** [comment]
I am clearly missing something.
the router should be the main resolver which is, and all in your config is ok. 
the hosts file is not the most elegant solution if you don't want to have static IPs and so on. 

In this case let me try to get some more data ...

### while connected to the VPN

can you ping the IP address? 
can I get the output of 
`nslookup <hostname> 192.168.0.254` or `dig <hostname> @192.168.0.254`
`nslookup <hostname> 10.8.0.1` or dig <hostname> @10.8.0.1`
`ns

**Comment by orazioedoardo** [comment]
Also `nmcli dev show | grep DNS` to check the upstream server used by Ubuntu.

**Comment by F0cks** [comment]
> I think for now is enough if I need more stuff will ask! =p

Wooaaaaw :D ! Ok let's go! 
## First try:
I have let `push "dhcp-option DNS 192.168.0.254"` in _server.conf_ and `192.168.0.2     test` in _hosts_.

> ### please reboot your raspberry pi and provide me the output of:
> uptime (sorry for this one but I really want to double check this one)
sudo systemctl status openvpn
sudo systemctl status openvpn@server

```
#####################################################################
pi@vpn:~ $ uptime
 13:19:45 up 0 min,  1 user,  load average: 1.69, 0.44, 0.15
#####################################################################
pi@vpn:~ $ sudo systemctl status openvpn
● openvpn.service - OpenVPN service
   Loaded: loaded (/lib/systemd/system/openvpn.service; enabled; vendor preset: 
   Active: active (exited) since Tue 2018-06-26 13:19:36 UTC; 42s ago
  Process: 355 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 355 (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/openvpn.service

Jun 26 13:19:36 vpn systemd[1]: Starting OpenVPN service...
Jun 26 13:19:36 vpn systemd[1]: Started OpenVPN service.
#####################################################################
pi@vpn:~ $ sudo systemctl status openvpn@server
● openvpn@server.service - OpenVPN connection to server
   Loaded: loaded (/lib/systemd/system/openvpn@.service; enabled; vendor preset:
   Active: active (running) since Tue 2018-06-26 13:19:28 UTC; 1min 7s ago
     Docs: man:openvpn(8)
           https://community.openvpn.net/openvpn/wiki/Openvpn23ManPage
           https://community.openvpn.net/openvpn/wiki/HOWTO
  Process: 247 ExecStart=/usr/sbin/openvpn --daemon ovpn-server --status /run/op
 Main PID: 258 (openvpn)
   CGroup: /system.slice/system-openvpn.slice/openvpn@server.service
           └─258 /usr/sbin/openvpn --daemon ovpn-server --status /run/openvpn/se

[...] removed: public IP
#####################################################################
```
> ### From VPN client in the local network without being connected to the VPN
>can you ping router ip addresS?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.254 -c 1
PING 192.168.0.254 (192.168.0.254) 56(84) bytes of data.
64 bytes from 192.168.0.254: icmp_seq=1 ttl=64 time=3.78 ms

--- 192.168.0.254 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 3.785/3.785/3.785/0.000 ms
```
> can you ping target host by its IP address?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.2 -c 1
PING 192.168.0.2 (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=64 time=8.69 ms

--- 192.168.0.2 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 8.694/8.694/8.694/0.000 ms
```
> can you ping target host by its hostname?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping test.local -c 1
PING test.local (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=64 time=10.0 ms

--- test.local ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 10.057/10.057/10.057/0.000 ms
```
> can I get the output of: nslookup <hostname> 192.168.0.254

Of course:
```
f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.254
Server:		192.168.0.254
Address:	192.168.0.254#53

** server can't find test: NXDOMAIN
```
Strange result spotted ...

> ### From client while connected to the VPN and outside of local Network (example 4G connection)
> can you ping router ip address?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.254 -c 1
PING 192.168.0.254 (192.168.0.254) 56(84) bytes of data.
64 bytes from 192.168.0.254: icmp_seq=1 ttl=63 time=71.9 ms

--- 192.168.0.254 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 71.978/71.978/71.978/0.000 ms
```
> can you ping target host by its IP address?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.2 -c 1
PING 192.168.0.2 (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=63 time=92.8 ms

--- 192.168.0.2 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 92.877/92.877/92.877/0.000 ms
```
> can you ping PiVPN server to 10.8.0.1 and to its LAN IP address?

Yes and yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 10.8.0.1 -c 1 && ping 192.168.0.22 -c 1
PING 10.8.0.1 (10.8.0.1) 56(84) bytes of data.
64 bytes from 10.8.0.1: icmp_seq=1 ttl=64 time=418 ms

--- 10.8.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 418.636/418.636/418.636/0.000 ms
PING 192.168.0.22 (192.168.0.22) 56(84) bytes of data.
64 bytes from 192.168.0.22: icmp_seq=1 ttl=64 time=72.0 ms

--- 192.168.0.22 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 72.030/72.030/72.030/0.000 ms
```
> can I get the output of:
> nslookup <hostname> 192.168.0.254 or dig <hostname> @192.168.0.254
> nslookup <hostname> 10.8.0.1 or dig <hostname> @10.8.0.1
> nsookup <hostname> <PiLocalAddress> or dig <hostname> @<PiLocalAddress

Of course!
```
f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.254
Server:		192.168.0.254
Address:	192.168.0.254#53

** server can't find test: NXDOMAIN

f0cks@f0cks-Alienware-15-R2:~$ nslookup test 10.8.0.1
;; connection timed out; no servers could be reached

f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.22
;; connection timed out; no servers could be reached
```
Strange result spotted 2...

## Second try:
This time, I have replaced `push "dhcp-option DNS 192.168.0.254"` by _push "route 192.168.0.0 255.255.255.0"_

> ### please reboot your raspberry pi and provide me the output of:
> uptime (sorry for this one but I really want to double check this one)
sudo systemctl status openvpn
sudo systemctl status openvpn@server

```
#####################################################################
pi@vpn:~ $ uptime
 13:49:44 up 1 min,  1 user,  load average: 1.02, 0.52, 0.20
#####################################################################
pi@vpn:~ $ sudo systemctl status openvpn
● openvpn.service - OpenVPN service
   Loaded: loaded (/lib/systemd/system/openvpn.service; enabled; vendor preset: 
   Active: active (exited) since Tue 2018-06-26 13:48:31 UTC; 1min 20s ago
  Process: 353 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 353 (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/openvpn.service

Jun 26 13:48:30 vpn systemd[1]: Starting OpenVPN service...
Jun 26 13:48:31 vpn systemd[1]: Started OpenVPN service.
#####################################################################
pi@vpn:~ $ sudo systemctl status openvpn@server
● openvpn@server.service - OpenVPN connection to server
   Loaded: loaded (/lib/systemd/system/openvpn@.service; enabled; vendor preset:
   Active: active (running) since Tue 2018-06-26 13:48:21 UTC; 1min 36s ago
     Docs: man:openvpn(8)
           https://community.openvpn.net/openvpn/wiki/Openvpn23ManPage
           https://community.openvpn.net/openvpn/wiki/HOWTO
  Process: 208 ExecStart=/usr/sbin/openvpn --daemon ovpn-server --status /run/op
 Main PID: 224 (openvpn)
   CGroup: /system.slice/system-openvpn.slice/openvpn@server.service
           └─224 /usr/sbin/openvpn --daemon ovpn-server --status /run/openvpn/se

[...] removed: public IP
#####################################################################

```
> ### From VPN client in the local network without being connected to the VPN
>can you ping router ip addresS?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.254 -c 1
PING 192.168.0.254 (192.168.0.254) 56(84) bytes of data.
64 bytes from 192.168.0.254: icmp_seq=1 ttl=64 time=3.66 ms

--- 192.168.0.254 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 3.664/3.664/3.664/0.000 ms
```
> can you ping target host by its IP address?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.2 -c 1
PING 192.168.0.2 (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=64 time=10.6 ms

--- 192.168.0.2 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 10.644/10.644/10.644/0.000 ms
```
> can you ping target host by its hostname?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping test.local -c 1
PING test.local (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=64 time=8.59 ms

--- test.local ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 8.591/8.591/8.591/0.000 ms
```
> can I get the output of: nslookup <hostname> 192.168.0.254

Of course:
```
f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.254
Server:		192.168.0.254
Address:	192.168.0.254#53

** server can't find test: NXDOMAIN
```
Strange result spotted 3 ...

> ### From client while connected to the VPN and outside of local Network (example 4G connection)
> can you ping router ip address?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.254 -c 1
PING 192.168.0.254 (192.168.0.254) 56(84) bytes of data.
64 bytes from 192.168.0.254: icmp_seq=1 ttl=63 time=72.9 ms

--- 192.168.0.254 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 72.912/72.912/72.912/0.000 ms
```
> can you ping target host by its IP address?

Yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 192.168.0.2 -c 1
PING 192.168.0.2 (192.168.0.2) 56(84) bytes of data.
64 bytes from 192.168.0.2: icmp_seq=1 ttl=63 time=84.0 ms

--- 192.168.0.2 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 84.005/84.005/84.005/0.000 ms
```
> can you ping PiVPN server to 10.8.0.1 and to its LAN IP address?

Yes and yes!
```
f0cks@f0cks-Alienware-15-R2:~$ ping 10.8.0.1 -c 1 && ping 192.168.0.22 -c 1
PING 10.8.0.1 (10.8.0.1) 56(84) bytes of data.
64 bytes from 10.8.0.1: icmp_seq=1 ttl=64 time=80.2 ms

--- 10.8.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 80.261/80.261/80.261/0.000 ms
PING 192.168.0.22 (192.168.0.22) 56(84) bytes of data.
64 bytes from 192.168.0.22: icmp_seq=1 ttl=64 time=77.3 ms

--- 192.168.0.22 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 77.345/77.345/77.345/0.000 ms
```
> can I get the output of:
> nslookup <hostname> 192.168.0.254 or dig <hostname> @192.168.0.254
> nslookup <hostname> 10.8.0.1 or dig <hostname> @10.8.0.1
> nsookup <hostname> <PiLocalAddress> or dig <hostname> @<PiLocalAddress

Of course!
```
f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.254
Server:		192.168.0.254
Address:	192.168.0.254#53

** server can't find test: NXDOMAIN

f0cks@f0cks-Alienware-15-R2:~$ nslookup test 10.8.0.1
;; connection timed out; no servers could be reached

f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.22
;; connection timed out; no servers could be reached

```
Strange result spotted 4 ...

And finally, the massive part of @orazioedoardo : 
>Also nmcli dev show | grep DNS to check the upstream server used by Ubuntu (still when connected to the vpn).

```
f0cks@f0cks-Alienware-15-R2:~$ nmcli dev show | grep DNS
IP4.DNS[1]:                             172.20.10.1
```

Good luck with this post @cfcolaco  `:O

**Comment by cfcolaco** [comment]
> This time, I have replaced push "dhcp-option DNS 192.168.0.254" by push "route 192.168.0.0 255.255.255.0"

Hmm sorry, maybe i wasn't clear enough ... the ` push "dhcp-option DNS 192.168.0.254"` is not to be replaced by `push "route 192.168.0.0 255.255.255.0"` 

please keep  the `push "route 192.168.0.0 255.255.255.0"`
and **ADD** to the file the `push "route 192.168.0.0 255.255.255.0"`

and repeat that part please. don't worry if I don't get back to you until Saturday .. as I am slowly conquering some time to dedicate to this project again, so I will take a while to go through all. 

thanks for providing all the info.

**Comment by orazioedoardo** [comment]
I think Ubuntu may still be using system dns instead of the ones provided by the config.

First, make sure you have `push "dhcp-option DNS 192.168.0.254"` in the server config.

Then check whether you have the file `/etc/openvpn/update-resolv-conf` on your computer.

If you do, add this lines to the client config, before the inline certificates.

> script-security 2
up /etc/openvpn/update-resolv-conf
down /etc/openvpn/update-resolv-conf

Try to connect, possibly with `sudo openvpn yourconfig.ovpn` first, then post the content of `/etc/resolv.conf`.

You should have 192.168.0.254. If so, test pinging some hosts.

**Comment by cfcolaco** [comment]
Gone through it all ... and something is not right at all ... 

> ```
> f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.254
> Server:		192.168.0.254
> Address:	192.168.0.254#53
> 
> ** server can't find test: NXDOMAIN
> 
> ```

and 

> 
> ```
> f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.254
> Server:		192.168.0.254
> Address:	192.168.0.254#53
> 
> ** server can't find test: NXDOMAIN
> 
> f0cks@f0cks-Alienware-15-R2:~$ nslookup test 10.8.0.1
> ;; connection timed out; no servers could be reached
> 
> f0cks@f0cks-Alienware-15-R2:~$ nslookup test 192.168.0.22
> ;; connection timed out; no servers could be reached
> ```

is the only place to get some clues. 
And indeed it seems that the resolution is not coming form the router, having this said this is why you are not resolving local IPs ... and your RaspberryPI is not resolving at all trough the resolv.conf.

let's focus on solving this the right way. 

what router and os do you have on your router? 
all machines involved in the process are using Linux? 

now time to make a query without specifying the DNS server and see where are those resolutions coming from. ( could have recalled this earlier in the loong big query of stuff to do xD )

run `nslookup test.local` and paste here the output please. 


no worries about taking time, I do it because I like to do it, I just feel sorry for not having that much time now to provide you the best support for everyone! =/

**Comment by F0cks** [comment]
Hummmmm : 
```
f0cks@f0cks-Alienware-15-R2:~$ nslookup test.local
Server:		127.0.1.1
Address:	127.0.1.1#53

** server can't find test.local: NXDOMAIN

```
I am probably wrong but, how does zeroconf works? How does it spread hostname in the network? I suppose it does not need any DNS server. If I understand that correctly, the "computer" spreads itself its hostname over the network. So my raspberry will be aware of this hostname cause it is on the same network (192.168.0.0/24). But none of client will catch it (different server). So one solution would be to catch all hostnames known with my raspberry (pivpn), store them, and spread them on the vpn network. But I cannot detect hostnames registered in hosts file...
Maybe we do not take the problem correctly, maybe we don't need to find what DNS server is used (maybe none) but find how to spread hard coded hostnames to VPN network.
Am I wrong or It could be working?

**Comment by lzilioli** [comment]
The following resolved it for me. 
https://github.com/pivpn/pivpn/issues/549#issuecomment-401070717

> Then check whether you have the file `/etc/openvpn/update-resolv-conf` on your computer.
> 
> If you do, add this lines to the CLIENT config, before the inline certificates.
> 
> > script-security 2
> > up /etc/openvpn/update-resolv-conf
> > down /etc/openvpn/update-resolv-conf

**Comment by 4s3ti** [comment]
long time without activity, 
please take this to pivpn subreddit, and refer to this issue there, so we can track and help if its still needed. 
Kind Regards.

**Comment by reinisb** [comment]
@F0cks and @lzilioli - I have the same issue. Ping to local devices via IP address works, but via their hostnames does not. Did you manage to find a way to fix this?

**Comment by lzilioli** [comment]
@reinisb nope I gave up. It’s incredibly frustrating. I came to a similar conclusion as you did. I’ve considered writing a little mini dns server that sits on my local network and resolves .local host names to ip addresses using ping, and then set my custom dns server as the dns server for devices on the vpn. I **think** it’ll work but I haven’t gotten around to it. Please keep me posted if you make any leeway here.

**Comment by orazioedoardo** [comment]
@reinisb You are right, musticast packets do not leave the local network, so when you ping ` raspberrypi.local` it tries to ping the local raspberry pi, not the one sitting inside the remote lan. You could definitely get it to work using openvpn's `tap` mode (bridged mode), since it forwards Ethernet frames. Another way I'm thinking about is a special dns server on the pi that provides some translation from raspberrypi.vpn -> raspberrypi.local and then issue query via mdns on your behalf. Perhaps you could use dnsmasq or a tweaked pihole to do that, however I haven't time to try neither of those now.

**Comment by reinisb** [comment]
@lzilioli and @orazioedoardo - thanks a lot for your responses. Yes, `tap` mode (layer 2) would fix this, but unfortunately, that is not an option in my case, since I need to use the `tun` mode (layer 3). Moreover, I do not have the skillset to create some sort of multicast packet repeater, so I am giving up. I hope that someday someone from the raspberry pi community will come up with a simple and elegant solution to this problem.

**Comment by 4s3ti** [comment]
I still don't really get what is wrong here... As far as my knowledge goes, to ping local network devices all needed usually is the already deprecated option (but might be worth trying to enforce it):
`push "route 192.168.0.0 255.255.255.0"` with the usage of /etc/hosts as suggested here: https://github.com/pivpn/pivpn/wiki/FAQ#how-to-resolve-local-hostnames

it can take more than one hostname, for example: 

`192.168.1.1    foo    foo.lan   foo.local   foo.mydomain   foobar foobar.lan`

This way the pi should know how to find these hosts. 

Other than that, and what i believe to be the propper way, you can setup the router as your DNS server and the router should know how to handle it.
If it doesn't you should find an equivalent to `/etc/hosts` on your router also some routers won't resolve if they are not the ones assigning IP address to the device, so I recommend assigning a static lease on the router itself this way the devices are still setup as DHCP, but the router will assign always the same IP address thus able to resolve the hostnames and achieving the same principle as static ip setup at the device side. 

**if the goal is to ping devices inside the VPN Network (i.e 10.8.0.0/24)** Then i don't really know ...  the only option I can see right now is to assign static ips to your clients and use the /etc/hosts file should do the trick. 

`192.168.1.1`  foo.lan`
`10.8.0.0.5`    foo.vpn`

hope this helps =) 

Kind regards.

**Comment by reinisb** [comment]
@4s3ti I did two things at the Raspberry Pi OpenVPN server:
(1) At `/etc/openvpn/server.conf` I added the push route rule of my local network address and its subnet mask;
(2) At `/etc/hosts` I added the IP address and its desired hostname of a device located at my server's local network (not the VPN network 10.8.0.0/24, but the regular one). Now, when I establish a VPN tunnel to the Raspberry Pi OpenVPN server from my OpenVPN client, then pinging the device located at the server's local network by its IP works, but with hostname does not.

Since my OpenVPN client device is a MacBook, I added the IP address and its desired hostname of a device located at my server's local network to the `/etc/hosts` file at my MacBook and now when I establish a tunnel to my server, then I can ping the device located at the server's local network both by its IP and hostname. So I can confirm that this works as a workaround.

That being said, this does not resolve the core limitation of multicast packets not being passed through a via VPN `tun` tunnel. Why would this be useful? Well, for example, at macOS a bonjour samba drive announces itself and shows up right at the Finder's sidebar under Locations. Thus there is no need to enter IP or hostname of that samba drive to access it.

**Comment by begleysm** [comment]
I had this problem for a long time.  And everyone told me to add `dhcp-options` to my server config (which I did) but it never worked.

I found that my `resolv.conf` file on my client wasn't updating with the `dhcp-options` pushed from the server.

I ultimately fixed everything by having the client run `resolveconf` which requires that `openvpn` be run with `--security-scripts 2`.

I wrote up specific, step by step instructions here: https://steamforge.net/wiki/index.php/How_to_configure_OpenVPN_to_resolve_local_DNS_%26_hostnames

Hope this helps.
